os_vif/objects/vif.py

```diff
@@ -45,96 +20,162 @@ class VIF(base.VersionedObject):
-        return phy_network
-
-    @property
-    def profileid(self):
-        return self.details.get(VIF_DETAILS_PROFILEID)
-
-    @property
-    def vlan(self):
-        return self.details.get(VIF_DETAILS_VLAN)
-
-    @property
-    def vhostuser_mode(self):
-        return self.details.get(VIF_DETAILS_VHOSTUSER_MODE)
-
-    @property
-    def vhostuser_socket(self):
-        return self.details.get(VIF_DETAILS_VHOSTUSER_SOCKET)
-
-    @property
-    def vhostuser_ovs_plug(self):
-        return self.details.get(VIF_DETAILS_VHOSTUSER_OVS_PLUG)
-
-    @property
-    def fixed_ips(self):
-        return [fixed_ip for subnet in self.network['subnets']
-                         for fixed_ip in subnet['ips']]
-
-    @property
-    def floating_ips(self):
-        return [floating_ip for fixed_ip in self.fixed_ips
-                            for floating_ip in fixed_ip['floating_ips']]
+
+class VIFBridge(VIFTap):
+    # Maps to libvirt type='bridge'
+
+    fields = {
+        # Name of the bridge device to connect to
+        'brname': fields.StringField(),
+
+        # Whether the bridge device should be automatically
+        # created if not already present
+        'autocreate': fields.BooleanField(),
+    }
+
+
+class VIFOpenVSwitch(VIFBridge):
+    # Maps to libvirt type='bridge' with openvswitch port profile
+
+    fields = {
+        'profileID': fields.StringField(),
+
+        'interfaceID': fields.UUIDField(nullable=True)
+    }
+
+
+class VIFMidonet(VIFBridge):
+    # Maps to libvirt type='bridge' with midonet port profile
+
+    fields = {
+        'interfaceID': fields.UUIDField(nullable=True)
+    }
+
+
+class VIFHostDevice(VIF):
+    # Abstract base
+
+    fields = {
+        # The PCI address of the device to assign
+        # TODO(berrange): should be a fields.PCIAddressField()
+        'address': fields.StringField()
+
+        # The VLAN number to configure
+        'vlan': fields.IntegerField()
+    }
+
+
+class VIFHostDeviceDirect(VIFHostDevice):
+    # Maps to libvirt type='direct' with mode='passthrough'
+    pass
+
+
+class VIFHostDeviceAssigned(VIFHostDevice):
+    # Maps to libvirt hostdev (ie PCI device assignment)
+    pass
+
+
+class VIFDirect(VIF):
+    # Abstract base
+
+    fields = {
+        # Name of the ethernet device to associate with
+        'devname': fields.StringField()
+    }
+
+
+class VIFDirect801QBG(VIFDirect):
+    fields = {
+        'managerID': fields.IntegerField(nullable=True),
+
+        'typeID': fields.IntegerField(nullable=True),
+
+        'typeIDVersion': fields.IntegerField(nullable=True),
+
+        'instanceID': fields.UUIDField(nullable=True),
+    }
+
+
+class VIFDirect801QBH(VIFDirect):
+
+    fields = {
+        'profileID': fields.StringField(),
+    }
+
+
+class VIFVHostUser(VIF):
+
+    fields = {
+        # Path to the UNIX domain socket
+        'path': fields.StringField()
+
+        # Sharing mode for the socket
+        'mode': fields.StringField()
+    }
```

Comment:
Yeah, I like what you've done above with the derived classes. ++

Only thing to correct would be the camelCasing of the attribute names :)